### PR TITLE
Improve schedule load, add slice_like

### PR DESCRIPTION
**Pull request by kevinthesun**
1. Enhance x86 conv2d schedule loading to allow more flexible schedule loading patten.
2. Add crop_like operator.

@yzhliu

### CODE REVIEW

**Comment by srkreddy1238** [comment]
@kevinthesun 

"Croping only 2nd and 3rd dimensions" and "Always crop based on other tensor shape" look like a special case consideration to me.

I suggest below changes for crop/crop_like for consideration to keep it more generalized.

- TVM/TOPI interface could be (Input Tensor, Array of Offsets, Array Of Lengths) to keep it more independent.
- At NNVM/TOP you could add logic to derive Lengths and Offsets based on inputs and attributes.
- Along with crop_like you could add crop to nnvm which doesn't do anything but call tvm/topi.

Please await for others opinion ..

**Comment by yzhliu** [comment]
good to me for # 1 - Enhance x86 conv2d schedule loading to allow more flexible schedule loading pattern.

**Comment by tqchen** [comment]
reviewers, please try to make inline comments, or approve the change https://docs.tvm.ai/contribute/code_review.html#approve-and-request-changes-explicitly

File: nnvm/include/nnvm/top/tensor.h [hunk 1]
```diff
@@ -259,6 +259,17 @@ struct ClipParam : public dmlc::Parameter<ClipParam> {
   }
 };
 
+struct CropLikeParam : public dmlc::Parameter<CropLikeParam> {
+  Tuple<int> offsets;
+  bool center_crop;
```
Comment (srkreddy1238) [inline nnvm/include/nnvm/top/tensor.h]: Handle the layout as well.

File: nnvm/src/top/tensor/transform.cc [hunk 2]
```diff
@@ -879,5 +879,48 @@ Examples::
...
+                          std::vector<TShape>* out_attrs) {
+  CHECK_EQ(in_attrs->size(), 2U);
+  CHECK_EQ(out_attrs->size(), 1U);
+  NNVM_ASSIGN_OUTPUT_SHAPE(attrs, *out_attrs, 0, (*in_attrs)[1]);
+  return true;
+}
+
+NNVM_REGISTER_OP(crop_like)
+.describe(R"code(Crop the 2nd and 3rd dim of input data,
+with width and height of the second input symbol.
```
Comment (srkreddy1238) [inline nnvm/src/top/tensor/transform.cc]: A new operator to call tvm/crop directly

File: topi/include/topi/transform.h [hunk 3]
```diff
@@ -563,5 +563,52 @@ inline Tensor take(const Tensor& a,
...
+* of h_w or with width and height of the second input tensor.
+*
+* \param inputs The input tensors.
+* \param offset Crop offset coordinate.
+* \param h_w Crop height and width. Only used when there is only one input tensors.
+* \param center_crop Whether to center crop.
+* \param name The name of the operation.
+* \param tag The tag to mark the operation.
+*
+* \return Cropped tensor
```
Comment (srkreddy1238) [inline topi/include/topi/transform.h]: TVM/TOPI interface could be (Input Tensor, Array of Offsets, Array Of Lengths) to keep it more independent.

**Comment by tqchen** [comment]
@kevinthesun can you act or make updated rely on the review comments?

File: nnvm/include/nnvm/top/tensor.h (hunk 1 above)
Comment (kevinthesun) [reply nnvm/include/nnvm/top/tensor.h]: The reason I didn't add layout as parameter is this operator's behavior is unrelated to layout. It just crops on the second and third axis.

File: nnvm/src/top/tensor/transform.cc (hunk 2 above)
Comment (kevinthesun) [reply nnvm/src/top/tensor/transform.cc]: Issue for a single crop operator in nnvm is that number of inputs can be either 1 or 2. It requires an extra parameter to hint whether it is a "crop_like" parameter. I can add another crop operator in nnvm which just takes 1 input tensor.

File: topi/include/topi/transform.h (hunk 3 above)
Comment (kevinthesun) [reply topi/include/topi/transform.h]: Is this related to the doc string?

**Comment by tqchen** [comment]
I am less sure about introducing crop operator, as crop seems to be a special case of slice operator( #1165), if shape inference can be done on the fly during conversion, it is likely we won't need the crop operator at all. 

@kevinthesun maybe you can remove the crop related stuffs first and merge the rest part it?

**Comment by kevinthesun** [comment]
Yes. Slice is more general. I'll remove crop from topi.

**Comment by tqchen** [comment]
@srkreddy1238 please review again on the updated changes

File: nnvm/src/top/tensor/transform.cc [hunk 4]
```diff
@@ -879,5 +879,91 @@ Examples::
...
+  CHECK_EQ(out_attrs->size(), 1U);
+  const SliceLikeParam& param = nnvm::get<SliceLikeParam>(attrs.parsed);
+  const TShape& src_shape = in_attrs->at(0);
+  const TShape& target_shape = in_attrs->at(1);
+  Tuple<dim_t> end_idx;
+  end_idx = Tuple<dim_t>(src_shape);
+  if (param.axis.ndim() == 0) {
+    for (size_t i = 0; i < src_shape.ndim(); ++i) {
+      if (i < target_shape.ndim()) {
+        end_idx[i] = target_shape[i];
```
Comment (srkreddy1238) [inline nnvm/src/top/tensor/transform.cc]: Add a check to make sure the end_idx doesn't go beyond input length.

File: nnvm/src/top/tensor/transform.cc [hunk 5]
```diff
@@ -879,5 +879,91 @@ Examples::
...
+    }
+  } else {
+    for (auto i : param.axis) {
+      if (i < 0) {
+        i = src_shape.ndim() + i;
+      }
+      CHECK_LT(i, target_shape.ndim())
+        << "Axis " << i << " exceeds dimension "
+        << target_shape.ndim()<< " of target_shape.";
+      end_idx[i] = target_shape[i];
```
Comment (srkreddy1238) [inline nnvm/src/top/tensor/transform.cc]: Same check as above  here as well.

File: nnvm/src/top/tensor/transform.cc [hunk 6]
```diff
@@ -879,5 +879,91 @@ Examples::
...
+  TShape out_shape = TShape(std::move(end_idx));
+  NNVM_ASSIGN_OUTPUT_SHAPE(attrs, *out_attrs, 0, out_shape);
+  return true;
+}
+
+NNVM_REGISTER_OP(slice_like)
+.describe(R"code(Slice the first input respect to the second input.
+)code" NNVM_ADD_FILELINE)
+.add_argument("data", "Tensor", "Input data to be sliced.")
+.add_argument("shape_like", "Tensor", "Shape like tensor")
```
Comment (srkreddy1238) [inline nnvm/src/top/tensor/transform.cc]: shape_like or slice_like ?

File: nnvm/src/top/tensor/transform.cc [hunk 7]
```diff
@@ -879,5 +879,91 @@ Examples::
...
+.set_attr<FCorrectLayout>("FCorrectLayout", ElemwiseBinaryKeepLeftLayout)
+.set_attr<FTVMCompute>(
+  "FTVMCompute", [](const NodeAttrs& attrs,
+                    const Array<Tensor>& inputs,
+                    const Array<Tensor>& out_info) {
+    const auto& param = nnvm::get<SliceLikeParam>(attrs.parsed);
+    Array<Expr> src_shape = inputs[0]->shape;
+    Array<Expr> target_shape = inputs[1]->shape;
+    Array<Expr> begin_idx, end_idx, strides;
+    for (size_t i = 0; i < src_shape.size(); ++i) {
```
Comment (srkreddy1238) [inline nnvm/src/top/tensor/transform.cc]: Merge these two loops.

File: nnvm/src/top/tensor/transform.cc [hunk 8]
```diff
@@ -879,5 +879,91 @@ Examples::
...
+        if (i < target_shape.size()) {
+          end_idx.Set(i, target_shape[i]);
+        }
+      }
+    } else {
+      for (int axis : param.axis) {
+        if (axis < 0) {
+          axis = static_cast<int>(src_shape.size()) + axis;
+        }
+        end_idx.Set(static_cast<size_t>(axis), target_shape[axis]);
```
Comment (srkreddy1238) [inline nnvm/src/top/tensor/transform.cc]: Same check as above  here as well.

File: nnvm/src/top/tensor/transform.cc [hunk 9]
```diff
@@ -879,5 +879,91 @@ Examples::
...
+      begin_idx.push_back(make_const(tvm::Int(32), 0));
+    }
+    for (size_t i = 0; i < src_shape.size(); ++i) {
+      strides.push_back(make_const(tvm::Int(32), 1));
+    }
+    end_idx = Array<Expr>(src_shape);
+    if (param.axis.ndim() == 0) {
+      for (size_t i = 0; i < src_shape.size(); ++i) {
+        if (i < target_shape.size()) {
+          end_idx.Set(i, target_shape[i]);
```
Comment (srkreddy1238) [inline nnvm/src/top/tensor/transform.cc]: Same check as above  here as well.

File: nnvm/src/top/tensor/transform.cc [hunk 10]
```diff
@@ -879,5 +879,91 @@ Examples::
...
+          axis = static_cast<int>(src_shape.size()) + axis;
+        }
+        end_idx.Set(static_cast<size_t>(axis), target_shape[axis]);
+      }
+    }
+    return Array<Tensor>{
+      topi::strided_slice(inputs[0], begin_idx, end_idx, strides)
+    };
+})
+.set_attr<FListInputNames>("FListInputNames", [](const NodeAttrs& attrs) {
```
Comment (srkreddy1238) [inline nnvm/src/top/tensor/transform.cc]: shape_like or slice_like ?

File: nnvm/tests/python/compiler/test_top_level4.py [hunk 11]
```diff
@@ -468,6 +468,57 @@ def test_nms():
...
+                i = len(np_data.shape) + i
+            end_idx[i] = np_shape_like.shape[i]
+    else:
+        for i in range(len(np_data.shape)):
+            if i < len(np_shape_like.shape):
+                end_idx[i] = np_shape_like.shape[i]
+    slice_idx = []
+    for b, e in zip(begin_idx, end_idx):
+        slice_idx.append(slice(b, e))
+    np_result = np_data[slice_idx]
```
Comment (srkreddy1238) [inline nnvm/tests/python/compiler/test_top_level4.py]: Make this numpy slicing as a function.

**Pull request merged**